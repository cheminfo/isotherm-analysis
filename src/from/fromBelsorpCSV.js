import { Analysis } from 'common-spectrum';

import { parseBelsorpCSV } from './parseBelsorpCSV';

export function fromBelsorpCSV(binary) {
  const result = parseBelsorpCSV(binary);
  const analysis = new Analysis();

  analysis.pushSpectrum(
    {
      x: result.adsorption['p/p0'],
      y: result.adsorption.na || result.adsorption.Va,
      p: result.adsorption.p0,
    },
    {
      dataType: 'Adsorption Isotherm',
      title: result.meta.COMMENT1,
      meta: result.meta,
    },
  );

  analysis.pushSpectrum(
    {
      x: result.desorption['p/p0'],
      y: result.desorption.na || result.adsorption.Va,
      p: result.desorption.p0,
    },
    {
      dataType: 'Desorption Isotherm',
      title: result.meta.COMMENT1,
      meta: result.meta,
    },
  );

  return analysis;
}
