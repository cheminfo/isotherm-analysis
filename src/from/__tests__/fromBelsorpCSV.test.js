import { readFileSync } from 'fs';
import { join } from 'path';

import { fromBelsorpCSV } from '../fromBelsorpCSV';

describe('fromBelsorpCSV', () => {
  it('belsorp.csv', () => {
    const binary = readFileSync(join(__dirname, 'data/belsorp.csv'));
    const analysis = fromBelsorpCSV(binary);
    expect(analysis.spectra).toMatchSnapshot();
  });
  it('belsorp2.csv', () => {
    const binary = readFileSync(join(__dirname, 'data/belsorp2.csv'));
    const analysis = fromBelsorpCSV(binary);
    expect(analysis.spectra).toMatchSnapshot();
  });
  it('belsorp3.csv', () => {
    const binary = readFileSync(join(__dirname, 'data/belsorp3.csv'));
    const analysis = fromBelsorpCSV(binary);
    expect(analysis.spectra).toMatchSnapshot();
  });
});
