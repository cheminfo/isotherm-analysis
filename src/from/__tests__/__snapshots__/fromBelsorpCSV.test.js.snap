// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromBelsorpCSV belsorp.csv 1`] = `
Array [
  Object {
    "dataType": "Adsorption Isotherm",
    "meta": Object {
      "Adsorption temperature": "77.000 [K]",
      "Adsorptive": "N2",
      "Adsorptive cross section area": "0.1620 [nm^2]",
      "Apparatus temperature": "0.0000 [C]",
      "COMMENT1": "Standard 1",
      "COMMENT2": "dsun",
      "COMMENT3": "",
      "COMMENT4": ", Vacuum degree before measurement:1.111E-4Pa",
      "Date of measurement": "18/02/06",
      "Dead volume": "30.402 [cm^3]",
      "Equilibrium time": "0 [SEC]",
      "File Name": "Act Carbon 2018 A.DAT",
      "File name of wall adsorption": "",
      "Number of adsorption data point :": "48",
      "Number of desorption data point :": "34",
      "Sample weight": "8.1300E-02 [g]",
      "Saturated vapor pressure": "100.000 [kPa]",
      "Serial number": "00463",
      "Standard volume": "24.502 [cm^3]",
      "Time of measurement": "30:25:37",
      "Version": "",
      "Wall adsorption correction value1": "",
      "Wall adsorption correction value2": "",
    },
    "title": "Standard 1",
    "variables": Object {
      "p": Object {
        "data": Array [
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
        ],
        "isMonotone": true,
        "label": "Relative pressure",
        "max": 100,
        "min": 100,
        "units": "kPa",
        "variable": "p0",
      },
      "x": Object {
        "data": Array [
          0.00002519,
          0.000053514,
          0.00012073,
          0.00029799,
          0.0012932,
          0.0017461,
          0.0026076,
          0.0054841,
          0.0078345,
          0.010417,
          0.015887,
          0.021483,
          0.04261,
          0.058021,
          0.073734,
          0.089467,
          0.1105,
          0.1382,
          0.1682,
          0.2097,
          0.2616,
          0.3141,
          0.355,
          0.4004,
          0.4518,
          0.502,
          0.552,
          0.6016,
          0.6512,
          0.7004,
          0.7488,
          0.7965,
          0.8417,
          0.8512,
          0.8841,
          0.8946,
          0.8989,
          0.9184,
          0.927,
          0.9322,
          0.9382,
          0.9413,
          0.9436,
          0.945,
          0.9464,
          0.9485,
          0.9495,
          0.9502,
        ],
        "isMonotone": true,
        "label": "p/p0",
        "max": 0.9502,
        "min": 0.00002519,
        "units": "",
        "variable": "p/p0",
      },
      "y": Object {
        "data": Array [
          0.000022289,
          0.000044887,
          0.00010545,
          0.00025223,
          0.00038821,
          0.00040127,
          0.00041572,
          0.00043709,
          0.00044646,
          0.00045753,
          0.00046899,
          0.00047417,
          0.00048363,
          0.00048831,
          0.00049304,
          0.00049674,
          0.00050272,
          0.00051142,
          0.00052507,
          0.00055171,
          0.00060275,
          0.00066873,
          0.00059409,
          0.00062153,
          0.00064272,
          0.0006607,
          0.00068386,
          0.00071687,
          0.00076635,
          0.00083787,
          0.0009399,
          0.0010829,
          0.0013107,
          0.0013923,
          0.0017455,
          0.0019372,
          0.0020646,
          0.0026912,
          0.0031789,
          0.003597,
          0.0042866,
          0.0048782,
          0.0054185,
          0.005928,
          0.006388,
          0.0072107,
          0.0079749,
          0.0087115,
        ],
        "isMonotone": false,
        "label": "Amount of absorbed gas",
        "max": 0.0087115,
        "min": 0.000022289,
        "units": "mol/g",
        "variable": "na",
      },
    },
  },
  Object {
    "dataType": "Desorption Isotherm",
    "meta": Object {
      "Adsorption temperature": "77.000 [K]",
      "Adsorptive": "N2",
      "Adsorptive cross section area": "0.1620 [nm^2]",
      "Apparatus temperature": "0.0000 [C]",
      "COMMENT1": "Standard 1",
      "COMMENT2": "dsun",
      "COMMENT3": "",
      "COMMENT4": ", Vacuum degree before measurement:1.111E-4Pa",
      "Date of measurement": "18/02/06",
      "Dead volume": "30.402 [cm^3]",
      "Equilibrium time": "0 [SEC]",
      "File Name": "Act Carbon 2018 A.DAT",
      "File name of wall adsorption": "",
      "Number of adsorption data point :": "48",
      "Number of desorption data point :": "34",
      "Sample weight": "8.1300E-02 [g]",
      "Saturated vapor pressure": "100.000 [kPa]",
      "Serial number": "00463",
      "Standard volume": "24.502 [cm^3]",
      "Time of measurement": "30:25:37",
      "Version": "",
      "Wall adsorption correction value1": "",
      "Wall adsorption correction value2": "",
    },
    "title": "Standard 1",
    "variables": Object {
      "p": Object {
        "data": Array [
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          undefined,
        ],
        "isMonotone": false,
        "label": "Relative pressure",
        "max": 100,
        "min": 100,
        "units": "kPa",
        "variable": "p0",
      },
      "x": Object {
        "data": Array [
          0.9453,
          0.9401,
          0.9355,
          0.9302,
          0.9242,
          0.9171,
          0.91,
          0.9023,
          0.881,
          0.8572,
          0.8491,
          0.8041,
          0.7963,
          0.7502,
          0.6992,
          0.6484,
          0.5981,
          0.5478,
          0.4981,
          0.4483,
          0.3985,
          0.3488,
          0.2991,
          0.2368,
          0.2064,
          0.1515,
          0.1229,
          0.099338,
          0.080461,
          0.066185,
          0.051949,
          0.037806,
          0.02102,
          0.014369,
          undefined,
        ],
        "isMonotone": true,
        "label": "p/p0",
        "max": 0.9453,
        "min": 0.014369,
        "units": "",
        "variable": "p/p0",
      },
      "y": Object {
        "data": Array [
          0.0082783,
          0.0072084,
          0.0062144,
          0.0052137,
          0.0043358,
          0.0036074,
          0.0029867,
          0.002568,
          0.0018215,
          0.0014289,
          0.00134,
          0.0010485,
          0.0010128,
          0.00086486,
          0.0007548,
          0.00068038,
          0.00063005,
          0.00060302,
          0.00058049,
          0.00056955,
          0.00054921,
          0.00051936,
          0.00048032,
          0.00043257,
          0.0004195,
          0.00041188,
          0.00041282,
          0.00041751,
          0.00042107,
          0.000423,
          0.00042577,
          0.00042812,
          0.00042641,
          0.00042234,
          undefined,
        ],
        "isMonotone": false,
        "label": "Amount of absorbed gas",
        "max": 0.0082783,
        "min": 0.00041188,
        "units": "mol/g",
        "variable": "na",
      },
    },
  },
]
`;

exports[`fromBelsorpCSV belsorp2.csv 1`] = `
Array [
  Object {
    "dataType": "Adsorption Isotherm",
    "meta": Object {
      "Adsorption temperature": "77 [K]",
      "Adsorptive": "N2",
      "Adsorptive cross section area": "0.162 [nm^2]",
      "Apparatus temperature": "0 [C]",
      "COMMENT1": "Standard 1",
      "COMMENT2": "dsun",
      "COMMENT3": "",
      "COMMENT4": ", Vacuum degree before measurement:1.111E-4Pa",
      "Date of measurement": "18.02.06",
      "Dead volume": "30.402 [cm^3]",
      "Equilibrium time": "0 [SEC]",
      "File Name": "Act Carbon 2018 A.DAT",
      "File name of wall adsorption": "",
      "Number of adsorption data point :": "48",
      "Number of desorption data point :": "34",
      "Sample weight": "8.13E-02 [g]",
      "Saturated vapor pressure": "100 [kPa]",
      "Serial number": "463",
      "Standard volume": "24.502 [cm^3]",
      "Time of measurement": "30:25:37",
      "Version": "",
      "Wall adsorption correction value1": "",
      "Wall adsorption correction value2": "",
    },
    "title": "Standard 1",
    "variables": Object {
      "p": Object {
        "data": Array [
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
        ],
        "isMonotone": true,
        "label": "Relative pressure",
        "max": 100,
        "min": 100,
        "units": "kPa",
        "variable": "p0",
      },
      "x": Object {
        "data": Array [
          0.0000252,
          0.0000535,
          0.000121,
          0.000298,
          0.00129,
          0.00175,
          0.00261,
          0.00548,
          0.00783,
          0.0104,
          0.0159,
          0.0215,
          0.0426,
          0.058,
          0.0737,
          0.0895,
          0.1105,
          0.1382,
          0.1682,
          0.2097,
          0.2616,
          0.3141,
          0.355,
          0.4004,
          0.4518,
          0.502,
          0.552,
          0.6016,
          0.6512,
          0.7004,
          0.7488,
          0.7965,
          0.8417,
          0.8512,
          0.8841,
          0.8946,
          0.8989,
          0.9184,
          0.927,
          0.9322,
          0.9382,
          0.9413,
          0.9436,
          0.945,
          0.9464,
          0.9485,
          0.9495,
          0.9502,
        ],
        "isMonotone": true,
        "label": "p/p0",
        "max": 0.9502,
        "min": 0.0000252,
        "units": "",
        "variable": "p/p0",
      },
      "y": Object {
        "data": Array [
          195.26,
          178.75,
          161.62,
          143.18,
          132.87,
          121.45,
          109.34,
          96.079,
          80.623,
          71.252,
          60.32,
          46.276,
          43.42,
          39.124,
          31.207,
          29.378,
          24.273,
          21.067,
          18.78,
          17.177,
          16.068,
          15.328,
          14.809,
          14.406,
          13.931,
          13.316,
          14.989,
          13.51,
          12.366,
          11.769,
          11.463,
          11.268,
          11.134,
          11.051,
          10.945,
          10.84,
          10.628,
          10.512,
          10.255,
          10.007,
          9.797,
          9.318,
          8.994,
          8.7014,
          5.6535,
          2.3636,
          1.0061,
          0.4996,
        ],
        "isMonotone": false,
        "label": "Amount of absorbed gas",
        "max": 195.26,
        "min": 0.4996,
        "units": "cm^3(STP)/g",
        "variable": "Va",
      },
    },
  },
  Object {
    "dataType": "Desorption Isotherm",
    "meta": Object {
      "Adsorption temperature": "77 [K]",
      "Adsorptive": "N2",
      "Adsorptive cross section area": "0.162 [nm^2]",
      "Apparatus temperature": "0 [C]",
      "COMMENT1": "Standard 1",
      "COMMENT2": "dsun",
      "COMMENT3": "",
      "COMMENT4": ", Vacuum degree before measurement:1.111E-4Pa",
      "Date of measurement": "18.02.06",
      "Dead volume": "30.402 [cm^3]",
      "Equilibrium time": "0 [SEC]",
      "File Name": "Act Carbon 2018 A.DAT",
      "File name of wall adsorption": "",
      "Number of adsorption data point :": "48",
      "Number of desorption data point :": "34",
      "Sample weight": "8.13E-02 [g]",
      "Saturated vapor pressure": "100 [kPa]",
      "Serial number": "463",
      "Standard volume": "24.502 [cm^3]",
      "Time of measurement": "30:25:37",
      "Version": "",
      "Wall adsorption correction value1": "",
      "Wall adsorption correction value2": "",
    },
    "title": "Standard 1",
    "variables": Object {
      "p": Object {
        "data": Array [
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
          100,
        ],
        "isMonotone": true,
        "label": "Relative pressure",
        "max": 100,
        "min": 100,
        "units": "kPa",
        "variable": "p0",
      },
      "x": Object {
        "data": Array [
          0.0144,
          0.021,
          0.0378,
          0.0519,
          0.0662,
          0.0805,
          0.0993,
          0.1229,
          0.1515,
          0.2064,
          0.2368,
          0.2991,
          0.3488,
          0.3985,
          0.4483,
          0.4981,
          0.5478,
          0.5981,
          0.6484,
          0.6992,
          0.7502,
          0.7963,
          0.8041,
          0.8491,
          0.8572,
          0.881,
          0.9023,
          0.91,
          0.9171,
          0.9242,
          0.9302,
          0.9355,
          0.9401,
          0.9453,
        ],
        "isMonotone": true,
        "label": "p/p0",
        "max": 0.9453,
        "min": 0.0144,
        "units": "",
        "variable": "p/p0",
      },
      "y": Object {
        "data": Array [
          195.26,
          178.75,
          161.62,
          143.18,
          132.87,
          121.45,
          109.34,
          96.079,
          80.623,
          71.252,
          60.32,
          46.276,
          43.42,
          39.124,
          31.207,
          29.378,
          24.273,
          21.067,
          18.78,
          17.177,
          16.068,
          15.328,
          14.809,
          14.406,
          13.931,
          13.316,
          14.989,
          13.51,
          12.366,
          11.769,
          11.463,
          11.268,
          11.134,
          11.051,
          10.945,
          10.84,
          10.628,
          10.512,
          10.255,
          10.007,
          9.797,
          9.318,
          8.994,
          8.7014,
          5.6535,
          2.3636,
          1.0061,
          0.4996,
        ],
        "isMonotone": false,
        "label": "Amount of absorbed gas",
        "max": 195.26,
        "min": 0.4996,
        "units": "cm^3(STP)/g",
        "variable": "Va",
      },
    },
  },
]
`;

exports[`fromBelsorpCSV belsorp3.csv 1`] = `
Array [
  Object {
    "dataType": "Adsorption Isotherm",
    "meta": Object {
      "Adsorption data point number": "18",
      "Adsorption temperature": "77.360 [K]",
      "Adsorptive": "N2",
      "Adsorptive molecular weight": "28.013",
      "Analysis date": "06/08/2021",
      "Analysis time": "10:02:06",
      "Comment1": "test measurement",
      "Comment2": "",
      "Comment3": "",
      "Comment4": ", Vacuum degree before measurement:3.446E-5Pa",
      "Data file": "TS-074_A-Port1.DAT",
      "Desorption data point number": "9",
      "Free space": "41.112 [cm^3]",
      "Manifold volume": "34.034 [cm^3]",
      "Molecular cross-sectional area": "0.1620 [nm^2]",
      "S/N": "182",
      "Sample mass": "0.0548 [g]",
      "Sample molecular weight": "0.0000",
      "Saturation vapor pressure": "96.436 [kPa]",
      "Specific surface area": "0.0000 [m^2 g^-1]",
    },
    "title": "",
    "variables": Object {
      "p": Object {
        "data": Array [
          97.737,
          97.098,
          96.731,
          96.594,
          96.507,
          96.512,
          96.489,
          96.441,
          96.492,
          96.42,
          96.429,
          96.378,
          96.407,
          96.375,
          96.353,
          96.352,
          96.283,
          96.27,
        ],
        "isMonotone": false,
        "label": "Relative pressure",
        "max": 97.737,
        "min": 96.27,
        "units": " kPa",
        "variable": "p0",
      },
      "x": Object {
        "data": Array [
          0.00029762,
          0.0021711,
          0.0058163,
          0.013711,
          0.019327,
          0.025799,
          0.043602,
          0.106,
          0.2734,
          0.3041,
          0.4451,
          0.5195,
          0.7366,
          0.8751,
          0.8813,
          0.9334,
          0.9474,
          0.9521,
        ],
        "isMonotone": true,
        "label": "p/p0",
        "max": 0.9521,
        "min": 0.00029762,
        "units": "",
        "variable": "p/p0",
      },
      "y": Object {
        "data": Array [
          5.7645,
          10.006,
          12.816,
          15.044,
          16.074,
          16.645,
          18.213,
          21.016,
          26.064,
          26.952,
          30.59,
          32.761,
          42.234,
          60.51,
          62.52,
          93.558,
          112.06,
          122.29,
        ],
        "isMonotone": true,
        "label": "Amount of absorbed gas",
        "max": 122.29,
        "min": 5.7645,
        "units": "cm^3(STP)/g",
        "variable": "Va",
      },
    },
  },
  Object {
    "dataType": "Desorption Isotherm",
    "meta": Object {
      "Adsorption data point number": "18",
      "Adsorption temperature": "77.360 [K]",
      "Adsorptive": "N2",
      "Adsorptive molecular weight": "28.013",
      "Analysis date": "06/08/2021",
      "Analysis time": "10:02:06",
      "Comment1": "test measurement",
      "Comment2": "",
      "Comment3": "",
      "Comment4": ", Vacuum degree before measurement:3.446E-5Pa",
      "Data file": "TS-074_A-Port1.DAT",
      "Desorption data point number": "9",
      "Free space": "41.112 [cm^3]",
      "Manifold volume": "34.034 [cm^3]",
      "Molecular cross-sectional area": "0.1620 [nm^2]",
      "S/N": "182",
      "Sample mass": "0.0548 [g]",
      "Sample molecular weight": "0.0000",
      "Saturation vapor pressure": "96.436 [kPa]",
      "Specific surface area": "0.0000 [m^2 g^-1]",
    },
    "title": "",
    "variables": Object {
      "p": Object {
        "data": Array [
          96.245,
          96.234,
          96.228,
          96.214,
          96.209,
          96.196,
          96.192,
          96.193,
          96.187,
          undefined,
        ],
        "isMonotone": false,
        "label": "Relative pressure",
        "max": 96.245,
        "min": 96.187,
        "units": " kPa",
        "variable": "p0",
      },
      "x": Object {
        "data": Array [
          0.9478,
          0.9185,
          0.8795,
          0.8032,
          0.6577,
          0.5122,
          0.3675,
          0.2637,
          0.1388,
          undefined,
        ],
        "isMonotone": true,
        "label": "p/p0",
        "max": 0.9478,
        "min": 0.1388,
        "units": "",
        "variable": "p/p0",
      },
      "y": Object {
        "data": Array [
          5.7645,
          10.006,
          12.816,
          15.044,
          16.074,
          16.645,
          18.213,
          21.016,
          26.064,
          26.952,
          30.59,
          32.761,
          42.234,
          60.51,
          62.52,
          93.558,
          112.06,
          122.29,
        ],
        "isMonotone": true,
        "label": "Amount of absorbed gas",
        "max": 122.29,
        "min": 5.7645,
        "units": "cm^3(STP)/g",
        "variable": "Va",
      },
    },
  },
]
`;
